riscv-smtdeleg-sstcfg.adoc

```diff
@@ -0,0 +1,98 @@
+[[intro]]
+== Introduction
+
+For a Machine-level environment, extension *Smtdeleg* (‘Sm’ for Privileged architecture and Machine-level extension, ‘tdeleg’ for Trigger Delegation) encompasses all added CSRs and all behavior modifications for a hart, over all privilege levels. For a
+Supervisor-level environment, extension *Sstcfg* (‘Ss’ for Privileged architecture and Supervisor-level extension, ‘tcfg’ for Trigger Configuration) provides access to delegated triggers. These extensions depend on the Sdtrig and Sscsrind extensions.
+
+[[body]]
+== CSRs
+
+=== Supervisor Trigger Select (`stselect`)
+
+`stselect` provides access to `tselect` from S-mode and VS-mode.
+
+== Supervisor and Virtual Supervisor Trigger Access
+
+While `siselect` holds XXX, the `sireg*` registers provide access to the `tdata*` and `tinfo` registers associated with the trigger selected in `tselect`.  The register mapping is shown below.
+
+.Indirect Trigger Register Mappings
+[width="50%",options="header"]
+|===
+| Indirect CSR | Sdtrig CSR 
+| `sireg` | `tdata1` 
+| `sireg2` | `tdata2` 
+| `sireg3` | `tdata3` 
+| `sireg4` | `tinfo` 
+|===
+
+[WARN]
+====
+_There has been much discussion of the best way to access counters in S-mode.  This method requires 1 new CSR (`stselect`), and creates a scenario where there are 2 levels of indirection (siselect routes sireg* to t*, and tselect routes t* to the state associated with the selected trigger)._
+
+_Also proposed was to simply allocate a range of `siselect` values that serve to select the trigger, bypassing `tselect` when accessing trigger state through `sireg*`.  This avoids the 2nd level of indirection, but artificially limits the number of triggers accessible in S-mode.  `tselect` supports up to 2^32^ triggers, while `siselect` would likely allocate only enough values to cover practical implementations (256?)._
+
+_A third approach was considered, akin to what is spec'd here but accessing `tselect` via `sireg5` instead of through a new `stselect` CSR.  While saving a CSR, this approach would require `sireg5` to continue to access `tselect` while the other `sireg*` registers shift the state they acccess based on the `tselect` value.  This approach seems more awkward for implementation._
+
+_Feedback on these options, or others, is welcome._
+====
+
+Similarly, when `vsiselect` holds XXX, the `vsireg*` registers provide access to `tdata*` and `tinfo` in the same manner as described above for `sireg*`.
+
+=== Trigger State Access Modifications
+
+`sireg*` and `vsireg*` provide access to a subset of the fields in the native trigger CSRs, and in some cases fields are relocated.  The field modifications that apply when trigger registers are accessed via `sireg*` or `vsireg*` are documented below.
+
+.Delegated Trigger CSR Field Modifications
+[options="header", cols="20%,30%,50%"]
+|===
+| Sdtrig Register | `sireg*` Field Modifications | `vsireg*` Field Modifications
+| `tdata1` | M is read-only 0 
+
+VS and VU are read-only 0
```

Comment:
S-mode should have the capability to set trigger for VS/VU mode. Is it pasted here by mistake?

Reply:
S-mode can enable triggers in VS/VU-mode via vsiselect/vsireg*.  This approach is much better for supporting nested virtualization, because it means a hypervisor running in VS-mode will naturally trap when it attempts to configure triggers for a true guest (by writing vsiselect/vsireg*), which will allow the L0 hypervisor (running in HS-mode) to emulate the desired behavior.

Reply:
Added a new non-normative comment to clarify this, see new text below the table.

Reply:
I agree with your statement on nested virtualization. Actually what I pointed out is that in `sireg*` column, VS and VU should be configurable instead of being "VS and VU are read-only 0". 
Moreover, it might be worthwhile to clairify  the relocation is applied to real CSR instead of programing ones.

Reply:
I'm not sure I follow.  HS-mode can enable trigger match in S/U-mode via siselect/sireg, and/or enable trigger match in VS/VU-mode via direct access to vsiselect/vsireg.  So it is intentional that VS and VU are ROZ in sireg, then are accessible only in vsireg.  I'm not sure what you mean by "real CSR instead of programming ones".  In case it's related to this: both siselect/sireg and vsiselect/vsireg are real physical CSRs, but from VS-mode only attempts to access siselect/sireg are redirected by the HW to vsiselect/vsireg.